Features/debian package

carrvo opened the pull request:
Re-create https://github.com/phokz/mod-auth-external/pull/62

Also, this work has a [mentors package](https://mentors.debian.net/package/libapache2-mod-authnz-external/)

carrvo committed:
added old changelog

carrvo committed:
added old copyright

carrvo committed:
other debian files from original package

carrvo committed:
fix source package name

carrvo committed:
fix source package name

carrvo committed:
fixes for `debuild -us -uc` to run

carrvo committed:
bring in more files from existing v3.3.2 deb package

carrvo committed:
modify for Apache2 specific helpers

based upon https://wiki.debian.org/Apache/PackagingFor24

carrvo committed:
include missing files to match existing package

carrvo committed:
cleanup debian/control

carrvo committed:
cleanup format

carrvo committed:
add debian packaging command to Makefile

carrvo committed:
fix timewalk

carrvo committed:
other commands used during package creation

carrvo committed:
check version before use

carrvo committed:
multi-line looks nicer

carrvo committed:
update URL and maintainers list

carrvo committed:
preparation for package-only

carrvo committed:
replace build-specific commands with download

carrvo committed:
build within downloaded package

carrvo committed:
remove unused tar command

carrvo committed:
ignore downloaded and built files

carrvo commented on debian/control:
```diff
@@ -0,0 +1,35 @@
+Source: libapache2-mod-authnz-external
+Maintainer: Josef Liška  <josef.liska@virtualmaster.com>
```
The uploader is not in the package's "Maintainer" or "Uploaders" fields

bimimicah replied:
### Original comment:
Josef Liška/phokz is the official head maintainer of mod_authnz_external, as designated by Jan Wolter's widow, to the best of my knowledge. Of course, as an open source project, anyone is welcome to fork it as long as they follow the license, but for the official repo, that is the correct attribution. 👍
<br>
### Additional comment:
I don't know what the Debian policy is, but maybe this field is supposed to be **you**, as the _source package_ maintainer, not the _source project_ maintainer?
You may want to check the [latest commit on the Debian package source repo](https://salsa.debian.org/debian/libapache2-mod-authnz-external/-/commit/ecb7f6ee6a08e84d0cda4e53ce7dc02731d1d107) from 3 months ago.

carrvo replied:
I have added myself to the `Uploaders` field so that the `Maintainers` can correctly attribute the source's maintainer. This may not be expected for Debian, but I like drawing awareness to the upstream and it resolves the warning.

carrvo commented on debian/control:
```diff
@@ -0,0 +1,35 @@
+Recommends: pwauth
+Homepage: https://github.com/phokz/mod-auth-external
+
+Package: libapache2-mod-authnz-external
```
E [malformed-deb-archive](https://udd.debian.org/lintian-tag/malformed-deb-archive)

    second (official) member control.tar.zst not control.tar.(gz|xz)
    third (official) member data.tar.zst not data.tar.(gz|xz|bz2|lzma)

carrvo replied:
This seems to be very strange and has been replicated for the same dev system for https://salsa.debian.org/debian/libapache2-mod-authnz-external/

Moving to an issue.

carrvo replied:
Moved to its own issue: https://github.com/carrvo/debian-libapache2-mod-authnz-external/issues/6

carrvo commented on debian/control:
```diff
@@ -0,0 +1,35 @@
+Installed-Size: 59
+Depends: ${shlibs:Depends}, ${misc:Depends}
+Recommends: pwauth
+Section: web
```
I [wrong-section-according-to-package-name](https://udd.debian.org/lintian-tag/wrong-section-according-to-package-name)

    web => httpd

carrvo replied:
Fixed.

carrvo commented on debian/copyright:
```diff
@@ -0,0 +1,67 @@
+Rai              <agentydragon@gmail.com>
+joakim-tjernlund
+
+It was originally located at http://code.google.com/p/mod-auth-external
```
W [obsolete-url-in-packaging](https://udd.debian.org/lintian-tag/obsolete-url-in-packaging)

    http://code.google.com/p/mod-auth-external [debian/copyright]

carrvo replied:
Ignoring.

carrvo commented on debian/changelog:
```diff
@@ -0,0 +1,25 @@
+libapache2-mod-authnz-external (3.3.2-2) unstable; urgency=medium
```
W [odd-historical-debian-changelog-version](https://udd.debian.org/lintian-tag/odd-historical-debian-changelog-version)

    3.3.2-2 (for native) [debian/changelog:9]

bimimicah replied:
There is a newer version (3.3.2-3) of this package that you might have missed, that should come in between 3.3.3 and 3.3.2-2. See:
[latest commit on the Debian package source repo from 3 months ago](https://salsa.debian.org/debian/libapache2-mod-authnz-external/-/commit/ecb7f6ee6a08e84d0cda4e53ce7dc02731d1d107)
[package update feed](https://tracker.debian.org/pkg/libapache2-mod-authnz-external/news/)
You may want to coordinate with [Petter Reinholdtsen](mailto:pere@hungry.com) or maybe he could help you get more familiar with Debian-related issues you come across?

carrvo replied:
This looks like literally everything I was looking for! (And vaguely like it is not orphaned despite Debian's listing...)

carrvo replied:
Contacting Debian package source to see if this repository is redundant.

carrvo replied:
Pulling in missing changelog entry from Debian package source.

carrvo replied:
Redundancy moved to its own issue: https://github.com/carrvo/debian-libapache2-mod-authnz-external/issues/7

carrvo commented on debian/control:
```diff
@@ -0,0 +1,35 @@
+Original-Maintainer: Debian QA Group <packages@qa.debian.org>
+Installed-Size: 59
+Depends: ${shlibs:Depends}, ${misc:Depends}
+Recommends: pwauth
+Section: web
+Priority: optional
+Homepage: https://github.com/phokz/mod-auth-external
```
I [installable-field-mirrors-source](https://udd.debian.org/lintian-tag/installable-field-mirrors-source)

    (in section for libapache2-mod-authnz-external) Homepage [debian/control:24]
    (in section for libapache2-mod-authnz-external) Priority [debian/control:23]
    (in section for libapache2-mod-authnz-external) Recommends [debian/control:21]
    (in section for libapache2-mod-authnz-external) Section [debian/control:22]

carrvo replied:
Fixed.

carrvo commented on debian/control:
```diff
@@ -0,0 +1,35 @@
+Maintainer: Josef Liška  <josef.liska@virtualmaster.com>
+Section: web
+Priority: optional
+Standards-Version: 4.7.0
```
I [out-of-date-standards-version](https://udd.debian.org/lintian-tag/out-of-date-standards-version)

    4.7.0 (released 2024-04-07) (current is 4.7.1)

carrvo replied:
Using highest dev version instead. This will actually be further outdated.

carrvo commented on debian/copyright:
P [no-dep5-copyright](https://udd.debian.org/lintian-tag/no-dep5-copyright)

    [debian/copyright]

carrvo replied:
Fixed.

carrvo commented on debian/control:
P [silent-on-rules-requiring-root](https://udd.debian.org/lintian-tag/silent-on-rules-requiring-root)

    [debian/control]

carrvo replied:
Ignoring.

carrvo commented on debian/README.Debian:
```diff
@@ -0,0 +1,17 @@
+README file
+http://blog.innerewut.de/2007/6/26/apache-2-2-authentication-with-mod_authnz_external
+http://www.unixpapa.com/mod_auth_external
+
```
P [trailing-whitespace](https://udd.debian.org/lintian-tag/trailing-whitespace)

    [debian/changelog:12]

carrvo commented on debian/control:
```diff
@@ -0,0 +1,35 @@
+Source: libapache2-mod-authnz-external
```
No VCS field present

carrvo replied:
Fixed.

carrvo reviewed (commented):
From https://mentors.debian.net/package/libapache2-mod-authnz-external/

carrvo committed:
first attempt at an mentors.debian.net package

carrvo commented on Makefile:
```diff
@@ -5,22 +5,28 @@ APXS=apxs
-	tar cvf mae.tar $(TAR)
+debian-package: libapache2-mod-authnz-external
+	cp -R debian libapache2-mod-authnz-external/
+	cd libapache2-mod-authnz-external && debuild --rootcmd=sudo --no-tgz-check -us -uc
```
Ideally elevating with sudo should be replaced with [instructions for debian/rules so that it builds to the appropriate directory](https://wiki.debian.org/Packaging/Intro?action=show&redirect=IntroDebianPackaging#Step_4:_Build_the_package).

bimimicah replied:
I'm not sure what debian's packaging typically looks like. On many distros, the mod_authnz_external package tends to be structured like this (I'm thinking specifically of Fedora/RedHat/Oracle/Rocky/etc. using RPM, since that's what I'm most familiar with):

1. [Source repo](https://src.fedoraproject.org/rpms/mod_authnz_external/tree/epel10): contain spec/script files for grabbing the project source from github/other source location, along with distro-specific files such as examples, configs, source patches, etc. Doesn't contain source, docs, examples, etc. for the actual project being packaged.
2. [Source package](https://kojipkgs.fedoraproject.org/packages/mod_authnz_external/3.3.3/11.el10_0/src/mod_authnz_external-3.3.3-11.el10_0.src.rpm): built from source repo, contains source repo files + actual project files from project repo
3. [per-architecture install package](https://kojipkgs.fedoraproject.org/packages/mod_authnz_external/3.3.3/11.el10_0/x86_64/mod_authnz_external-3.3.3-11.el10_0.x86_64.rpm): built for each CPU architecture from source package, contains just the final files to be added to the system + any installation scripts

### This part might be useful:
Running a build directly from our repo's standard makefile runs `apxs` to compile the module and "install"/copy it to the destination, but for a package build you probably just want to **tell `apxs` to only compile the module, not install it**.
Here is a [Fedora example](https://src.fedoraproject.org/rpms/mod_authnz_external/blob/epel10/f/mod_authnz_external.spec#_30) that may be helpful to you (see the **%build** section). Also see [`apxs` man page](https://man.freebsd.org/cgi/man.cgi?query=apxs) for help.

Hope this is helpful to you.

carrvo replied:
Additional information from the commentor on the suggestion is found in the newer `rules` file.
> 2. The original package repo has some files you probably will find useful that are missing/out-of-date here. The `rules` and `watch` files look like they are probably going to be pretty useful for you. (especially note the newer version's `rules` file)

carrvo replied:
I will move this work to another branch to better experiment and explore.

carrvo replied:
And own issue: https://github.com/carrvo/debian-libapache2-mod-authnz-external/issues/8

carrvo commented on debian/source/format:
```diff
@@ -0,0 +1 @@
+3.0 (native)
```
Should figure out how to use [quilt format](https://www.debian.org/doc/manuals/maint-guide/modify.en.html)

carrvo replied:
Moved to its own branch: https://github.com/carrvo/debian-libapache2-mod-authnz-external/tree/features/debian-package-quilt

carrvo replied:
And own issue: https://github.com/carrvo/debian-libapache2-mod-authnz-external/issues/9

bimimicah commented:
> This has been moved to [carrvo/debian-libapache2-mod-authnz-external#1](https://github.com/carrvo/debian-libapache2-mod-authnz-external/pull/1)

<br/>
I added some comments/review to the old PR as you requested. <s>Would you like me to add them to the new location?</s><br/>
I will go ahead and add them here.<br/>
<br/><br/>
Also, as mentioned in those comments, I've now split out the LICENSE file to make it easy to just ship it with the built package (to make it easier for you to satisfy Apache 1.0 license requirements)

carrvo committed:
DEP-5 copyright format

carrvo committed:
Vcs field

bimimicah commented on debian/README.Debian:
```diff
@@ -0,0 +1,17 @@
+This package can be used with pwauth to provide secure auth against PAM without
```
It would probably be good to include some basic explanation from the main project readme file at the top of this file, as well as a link to the project homepage. I guess the previous Debian maintainer probably only used it with `pwauth`, but these are actually 2 separate projects.

carrvo replied:
I will clean up this file. However, as per [the description of the file](https://www.debian.org/doc/manuals/maint-guide/dother.en.html#readme) I will not include anything from the main project README (but I will still distribute and refer to it).

carrvo replied:
Done.

bimimicah commented on debian/README.Debian:
```diff
@@ -0,0 +1,17 @@
+This package can be used with pwauth to provide secure auth against PAM without
+exposing /etc/shadow file.
+
+To use it on Debian, make sure to also pwauth package then add these lines to
```
probably should say:
"also `add the` pwauth package"
or
"also `install the` pwauth package"

carrvo replied:
Re-worked.

carrvo committed:
add info for package uploader

carrvo committed:
fix format issue

carrvo committed:
change section according to suggestion

carrvo committed:
remove warning fields

carrvo committed:
remove warning fields

carrvo committed:
Makefile with upload commands

bimimicah commented:
Well, I think that's all the comments I had before. Hope they are helpful to you.
**Honestly of everything I wrote the only thing that really matters is the license issue.** The old Apache License 1.0 specifically requires that it be distributed alongside the binary package, how it looks in the end (separate `LICENSE`/`CONTRIBUTORS` files or appended to the bottom of a `copyright` file) doesn't really matter, but the **text** of the `LICENSE` and `CONTRIBUTORS` files should end up somewhere on the user's machine associated with this package.
I figure since the package is being updated, now is a good time to get this done right. I will be contacting the Fedora/EPEL maintainers to get their package fixed as well.

2 more before I call it a night:

1. If the Debian policy is to keep original project source in your package repo, then just unzip to that folder when you have new source release. <b>If that is not the policy, then it would be good to get rid of the separate copies of the documentation files</b> (AUTHENTICATORS, TODO, UPGRADE, etc.) and pull them in when you grab the project source.
2. The original package repo has some files you probably will find useful that are missing/out-of-date here. The `rules` and `watch` files look like they are probably going to be pretty useful for you. (especially note the newer version's `rules` file)

carrvo commented:
@bimimicah this has been quite a helpful review! I really appreciate it :heart:

I especially appreciate the licensing note. That is something often easy to mix up or get wrong.

For the duplicate files, I intend to remove them after I ensure any important information has been copied over to any relevant place. Not strictly necessary to wait since the information lives upstream, but it works as a good reminder.

Hopefully I can resolve things by end of week!

carrvo committed:
distribute additional files from upstream

carrvo committed:
fix copyright to have license text in appropriate location

carrvo committed:
fix Vcs-Browser warning

carrvo committed:
use Standards-Version of dev system

carrvo committed:
re-work README.Debian

carrvo commented:
> Well, I think that's all the comments I had before. Hope they are helpful to you. **Honestly of everything I wrote the only thing that really matters is the license issue.** The old Apache License 1.0 specifically requires that it be distributed alongside the binary package, how it looks in the end (separate `LICENSE`/`CONTRIBUTORS` files or appended to the bottom of a `copyright` file) doesn't really matter, but the **text** of the `LICENSE` and `CONTRIBUTORS` files should end up somewhere on the user's machine associated with this package. I figure since the package is being updated, now is a good time to get this done right. I will be contacting the Fedora/EPEL maintainers to get their package fixed as well.

Since I am pulling the frozen v3.3.3 I will leave this in just the `debian/copyright` so that the text is there. Further releases I will distribute the separated file in addition to the `debian/copyright` (and perhaps simplify it).

> 1. If the Debian policy is to keep original project source in your package repo, then just unzip to that folder when you have new source release. If that is not the policy, then it would be good to get rid of the separate copies of the documentation files (AUTHENTICATORS, TODO, UPGRADE, etc.) and pull them in when you grab the project source.

This was a reminder that there is relevant information that would be worth copying to appropriate locations. I have now decided to distribute some of these files.

carrvo committed:
Pulling missing changelog entry from salsa.debian.org

https://salsa.debian.org/debian/libapache2-mod-authnz-external/-/commit/ecb7f6ee6a08e84d0cda4e53ce7dc02731d1d107

carrvo commented:
> > Well, I think that's all the comments I had before. Hope they are helpful to you. **Honestly of everything I wrote the only thing that really matters is the license issue.** The old Apache License 1.0 specifically requires that it be distributed alongside the binary package, how it looks in the end (separate `LICENSE`/`CONTRIBUTORS` files or appended to the bottom of a `copyright` file) doesn't really matter, but the **text** of the `LICENSE` and `CONTRIBUTORS` files should end up somewhere on the user's machine associated with this package. I figure since the package is being updated, now is a good time to get this done right. I will be contacting the Fedora/EPEL maintainers to get their package fixed as well.
> 
> Since I am pulling the frozen v3.3.3 I will leave this in just the `debian/copyright` so that the text is there. Further releases I will distribute the separated file in addition to the `debian/copyright` (and perhaps simplify it).
> 

Follow up: https://github.com/carrvo/debian-libapache2-mod-authnz-external/issues/10

bimimicah commented:
> Since I am pulling the frozen v3.3.3 I will leave this in just the `debian/copyright` so that the text is there. Further releases I will distribute the separated file in addition to the `debian/copyright` (and perhaps simplify it).

Good point. I should see if I can carve out some time to add testing infrastructure to the automated builds to make some progress on clearing the blockers for v3.3.4.

Please note the contents of `debian/copyright` are not currently correct, specifically:
1.  the last paragraph of the license is missing
2. the copyright header at the top is the generic Apache license copyright statement, not ours (from v.3.3.3) with the past maintainers and contributors (who are all joint copyright holders)
     &nbsp;    a. technically the old lines of code which are still under "Apache Group" copyright are a few boilerplate lines that have almost certainly been replaced during upgrades to work with newer versions of Apache, but the project as a whole is still a "derivative work" regardless. My point being they don't really hold the primary copyright and shouldn't be the only attribution in the copyright header.
3. If you feel the contributor list would go better in the comments section of the file, I suppose that makes no difference as long as the list ends up on the user's machine, but lines 58-73 really should be replaced with the contents of the v.3.3.3 `CONTRIBUTORS` file, formatted/indented however necessary for Debian standards.
4. The reference to Google Code is unnecessary as this repository replaces that one (as you can see from the Google Code project homepage).
5. The reference to the unixpapa site in the `readme` has the same issue. That is Jan Wolter's old site, archived when he passed away; it will never be updated and you cannot find support there.

Hope that clarifies things :)

carrvo commented:
> Please note the contents of `debian/copyright` are not currently correct, specifically:
> 
>     1. the last paragraph of the license is missing
> 
>     2. the copyright header at the top is the generic Apache license copyright statement, not ours (from v.3.3.3) with the past maintainers and contributors (who are all joint copyright holders)
>             a. technically the old lines of code which are still under "Apache Group" copyright are a few boilerplate lines that have almost certainly been replaced during upgrades to work with newer versions of Apache, but the project as a whole is still a "derivative work" regardless. My point being they don't really hold the primary copyright and shouldn't be the only attribution in the copyright header.
> 
>     3. If you feel the contributor list would go better in the comments section of the file, I suppose that makes no difference as long as the list ends up on the user's machine, but lines 58-73 really should be replaced with the contents of the v.3.3.3 `CONTRIBUTORS` file, formatted/indented however necessary for Debian standards.
> 
>     4. The reference to Google Code is unnecessary as this repository replaces that one (as you can see from the Google Code project homepage).
> 
>     5. The reference to the unixpapa site in the `readme` has the same issue. That is Jan Wolter's old site, archived when he passed away; it will never be updated and you cannot find support there.
> 

Created issue https://github.com/carrvo/debian-libapache2-mod-authnz-external/issues/14

carrvo commented:
> > Since I am pulling the frozen v3.3.3 I will leave this in just the `debian/copyright` so that the text is there. Further releases I will distribute the separated file in addition to the `debian/copyright` (and perhaps simplify it).
> 
> Good point. I should see if I can carve out some time to add testing infrastructure to the automated builds to make some progress on clearing the blockers for v3.3.4.
> 

@bimimicah  you may want to **consider** releasing it as `v3.4.0` or `v4.0.0` because the following (in `v3.3.3`) is a change to the API (and the change I am trying to consume) and a higher version increment can help bring that to attention:
> Added the "GroupExternalAuthNCheck Off" directive to disable the check
for a previously authenticated user when using only the group
authorization module. If you use any user authentication module in the
same context, this may be undesirable. Thanks to Micah Andersen/BIMI.

bimimicah commented:
> you may want to **consider** releasing it as `v3.4.0` or `v4.0.0` because the following (in `v3.3.3`) is a change to the API (and the change I am trying to consume) and a higher version increment can help bring that to attention:
> 
> > Added the "GroupExternalAuthNCheck Off" directive to disable the check
> > for a previously authenticated user when using only the group
> > authorization module. If you use any user authentication module in the
> > same context, this may be undesirable. Thanks to Micah Andersen/BIMI.

Hmmm. Interesting suggestion; I'll have to think about that. The big reason many distros haven't moved beyond the last few versions (3.3.0, 3.3.1, 3.3.2) is because this is a relatively stable project with few major changes; mostly small fixes.
So a lot of times, historically, I think people haven't moved to a newer version until they needed to upgrade Apache httpd and needed a newer version of this to go with it.
There hasn't been a major breaking change to Apache in over a decade (version 2.4 introduced changes which were resolved by our v3.3.0), so some distros haven't bothered to look for an updated mod_authnz_external package since then (OpenMandriva and NetBSD are good examples).
For similar reasons, ever since the 3.x series started in 2006, we basically don't make breaking changes until Apache does.

That's one of the big reasons I haven't made the v.3.3.4 release yet (see phokz/mod-auth-external#37). I have been worried because it has a pretty substantial addition (AUTHORIZE_ headers are added to the external processes' environment values, the first environment variable addition since CONTEXT was added in 2008) which could theoretically break applications that have been expecting a specific environment table size for the last 2 decades.... And now that I have that all in writing, I think I've made up my mind about phokz/mod-auth-external#37, I'll definitely put it behind a flag.
Basically all of that to say, our versioning system so far has been basically "Apache httpd-based", or something like this:
Read this table from the **bottom up**, it makes more sense that way....
```
Major.Minor.Revision
  ^     ^      ^
  |     |      |
  |     |      ➥ Any other changes
  |     ➥ New Apache minor versions based on the same design (e.g. when 3.3 fixed minor issues to upgrade us to Apache 2.4)
  |     ➥ Internal (non-breaking) rewrites (e.g. when 3.2 switched to use APR under the hood)
  ➥ Complete rewrites of internal and external design (e.g. when 3.0/3.1 introduced the Apache authnz model)
```

If we switched to semver ([Semantic Versioning](https://semver.org/)), which seems to be the standard nowadays, then it would look like this:
Read this table from the **top down**, it makes more sense that way....
```
Major.Minor.Revision
  ^     ^      ^
  |     |      |
  |     |      ➥ Non-breaking _bug-fixes_ only (e.g. when 3.3.3 fixed the parenthesis grouping bug that was unlikely to ever trigger)
  |     ➥ Non-breaking new features (e.g. when 3.3.3 added the GroupExternalAuthNCheck directive/flag)
  |     ➥ Non-breaking internal changes (e.g. when 3.2 switched to use APR under the hood)
  ➥ Anything that breaks compatibility, including:
          Complete rewrites of internal and external design (e.g. when 3.0/3.1 introduced the Apache authnz model)
          New Apache minor versions based on the same design (e.g. when 3.3 upgraded to Apache 2.4 and removed AuthBasicAuthoritative and GroupExternalAuthoritative)
```

**Essentially, if we switched to semver, then 3.3.4 would actually be 3.4.0 and if anyone (me, I guess) ever gets around to adding the features for phokz/mod-auth-external#40 and phokz/mod-auth-external#21, then those would be in 3.5.0 or later.**

bimimicah commented:
> @bimimicah you may want to **consider** releasing it as `v3.4.0` or `v4.0.0` because the following (in `v3.3.3`) is a change to the API (and the change I am trying to consume) and a higher version increment can help bring that to attention:
> 
> > Added the "GroupExternalAuthNCheck Off" directive to disable the check
> > for a previously authenticated user when using only the group
> > authorization module. If you use any user authentication module in the
> > same context, this may be undesirable. Thanks to Micah Andersen/BIMI.

I've added phokz/mod-auth-external#63 to address this. I think this is probably a good idea, unless someone has a reason why it's not.

carrvo commented:
I'll admit, in my currently short time as a software developer, semantic versioning is what I have generally encountered and worked with. Apparently I got too used to it to remember its name :frowning:
